include/pybind11/typing.h

```diff
@@ -0,0 +1,94 @@
+/*
+    pybind11/typing.h: Convenience wrapper classes for basic Python types
+    with more explicit annotations.
+
+    Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "detail/common.h"
+#include "cast.h"
+#include "pytypes.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+/*
+    The following types can be used to direct pybind11-generated docstrings
+    to have have more explicit types (e.g., `List[str]` instead of `list`).
+    Just use these in place of existing types.
+
+    There is no additional enforcement of types at runtime.
+*/
+
+template <typename... Types>
+class Tuple : public tuple {
+    using tuple::tuple;
+};
+
+template <typename K, typename V>
+class Dict : public dict {
+    using dict::dict;
+};
+
+template <typename T>
+class List : public list {
+    using list::list;
+};
+
+template <typename T>
+class Set : public set {
+    using set::set;
+};
+
+template <typename Signature>
+class Callable;
+
+template <typename Return, typename... Args>
+class Callable<Return(Args...)> : public function {
+    using function::function;
+};
+
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+template <typename... Types>
+struct handle_type_name<Tuple<Types...>> {
+    static constexpr auto name
+        = const_name("Tuple[") + concat(make_caster<Types>::name...) + const_name("]");
```

Comment:
Another thing I learned from the typing expert team mate:

>> Would it be better to have lower-case names (tuple, dict, list, set) in the output?

> It depends on what Python versions you need to support. In Python 3.10 and higher, it is indeed preferred to use the lowercase builtin names. 

That makes me think `#ifdef`s of some sort would be best (to produce `Tuple` for <= 3.9 and `tuple` otherwise), although that adds a bit of clutter until we can drop 3.9 support.

Tying that back to my other comment, `py::typing::tuple` vs `py::typing::Tuple`:

Pure Python `typing.Tuple` will likely be deleted at some point in the future, although IIUC there is no timeline yet.

We cannot reuse `py::tuple` like pure Python reuses `tuple` for typing purposes.

Should we have `py::typing::tuple` or `py::typing::Tuple`? I don't know. It's on the awkward side either way.

I think I'd settle for `py::typing::tuple` (for any Python version), to not look old fashioned in the long run.

Reply:
For Python 3.7-3.9, you still can use lower case as long as you only use them for typing. In this case, it can only be used for typing, not runtime, so please use the lowercase, un-namespaced names, no ifdefs. I've mostly removed typing.* stuff from libraries that are 3.7+.

Reply:
+1

Note that we have already firmly decided to remove Python 3.6 support asap (as soon as I find a moment).

Reply:
(Yes, rather assuming that in this discussion :) - besides, typing in 3.6 is pretty hard since everything has dropped it a while back)

Reply:
I would add a `class Union<T...> : object` to implement union markers in a more general way.

If there was a `py::typing` namespace it would make more sense to implement such a thing, since `py::Union` is a bit weird.

Reply:
> I would add a `class Union<T...> : object` to implement union markers in a more general way.
> 
> If there was a `py::typing` namespace it would make more sense to implement such a thing, since `py::Union` is a bit weird.

Sounds great to me. I felt strongly already that `py::typing` is the way to go, now even more.

Reply:
I gave it a shot, but I think adding  would be out of scope for this current PR because of two messy issues that came up.

If you add a `Union` type then you get compilation errors when loading the argument because one would need to define a `PYBIND11_OBJECT_COMMON` for it... which, you could add a check function to check if the incoming PyObject is one of the arg types, but the common types also have implicit conversions which wouldn't be caught here. 

Additionally this would require defining a `concat_pipe` version of `concat` which is mostly copy/paste, but it's another thing to maintain.